src/commands/templates.rs

```diff
@@ -252,6 +255,20 @@ impl Upgrade {
             return Ok(None);
         }
 
+        if !no_origin.is_empty() {
+            eprintln!(
+                "Spin could not determine where the following templates were installed from:"
+            );
+            for template in no_origin {
+                eprintln!("- {}", template.id());
+            }
+            eprintln!("To upgrade them, run `spin templates install --upgrade` with the --git or --dir option");
+            eprintln!();
+            if !self.all {
```

Comment:
is there a possibility of this message ("The following template repositories can be automatically upgraded:") printing and then there not being any with an origin to display after?

Reply:
I don't believe so.  This is ensured by the `sources.is_empty()` block starting on line 244 - it is guaranteed to early exit on 255.

I have no qualms about adding a belt and braces check here, though, if you think it's worth it?

(Well no qualms except the THIRD RUN through CI but okay that's not so much a qualm as a pet peeve.)

Reply:
Thats perfect as it is. Didnt take a deep look, so thanks for pointing out the check.